python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,855 @@
+
+        worker_id = self._sg_data_retriever.execute_nav_expand(
+            path, seed_entity_field, entity_fields)
+
+        # keep a lookup to map the worker id with the url it is querying
+        self._running_query_lookup[worker_id] = path
+
+    def _refresh_data(self):
+        """
+        Rebuild the data in the model to ensure it is up to date.
+
+        This call should be asynchronous and return instantly. The update should
+        be applied as soon as the data from Shotgun is returned.
+
+        If the model is empty (no cached data) no data will be shown at first
+        while the model fetches data from Shotgun.
+
+        As soon as a local cache exists, data is shown straight away and the
+        shotgun update happens silently in the background.
+
+        If data has been added, this will be injected into the existing
+        structure. In this case, the rest of the model is intact, meaning that
+        also selections and other view related states are unaffected.
+
+        If data has been modified or deleted, a full rebuild is issued, meaning
+        that all existing items from the model are removed. This does affect
+        view related states such as selection.
+        """
+
+        # get a list of all urls to update. these will be paths for all existing
+        # items that are not empty or have no children already queried. we know
+        # we always need to refresh the inital path.
+        urls = [self._path]
+        urls.extend(self._get_queried_urls_r(self.invisibleRootItem()))
+
+        # query in order of length
+        for url in sorted(set(urls)):
+            logger.debug("Refreshing hierarchy model url: %s" % (url,))
+            self._query_hierarchy(
+                url,
+                self._seed_entity_field,
+                self._entity_fields
+            )
+
+    def _update_item(self, item, data):
+        """
+        Updates the supplied item with the newly queried data.
+
+        :param item: A :class:`~PySide.QtGui.QStandardItem` instance to update.
+        :param dict data: The newly queried data.
+
+        :return: ``True`` if the item was updated, ``False`` otherwise.
+        """
+
+        # get a copy of the data and remove the child item info so that
+        # each item in the tree only stores data about itself
+        new_item_data = copy.deepcopy(data)
+        if "children" in data.keys():
+            del new_item_data["children"]
+
+        # compare with the item's existing data
+        old_item_data = get_sg_data(item)
+        if self._sg_compare_data(old_item_data, new_item_data):
+            # data has not changed
+            return False
+
+        # data differs. set the new data
+        item.setData(sanitize_for_qt_model(new_item_data), self.SG_DATA_ROLE)
+
+        # ensure the label is updated
+        item.setText(data["label"])
+
+        return True
+
+    def _update_subtree(self, item, nav_data):
```

Comment:
private? or add example to doc string of when it would be useful to subclass it.